xds/internal/testutils/e2e/clientresources.go

```diff
@@ -30,37 +32,74 @@ import (
+	ServerSideCertProviderInstance = "server-side-certificate-provider-instance"
+)
+
+// SecurityLevel allows the test to control the security level to be used in the
+// resource returned by this package.
+type SecurityLevel int
+
+const (
+	// SecurityLevelNone is used when no security configuration is required.
+	SecurityLevelNone SecurityLevel = iota
+	// SecurityLevelTLS is used when security configuration corresponding to TLS
+	// is required. Only the server presents an identity certificate in this
+	// configuration.
+	SecurityLevelTLS
+	// SecurityLevelMTLS is used when security ocnfiguration corresponding to
+	// mTLS is required. Both client and server present identity certificates in
+	// this configuration.
+	SecurityLevelMTLS
+)
+
+// ResourceParams wraps the arguments to be passed to DefaultClientResources.
+type ResourceParams struct {
+	// DialTarget is the client's dial target. This is used as the name of the
+	// Listener resource.
+	DialTarget string
+	// NodeID is the id of the xdsClient to which this update is to be pushed.
+	NodeID string
+	// Host is the host of the default Endpoint resource.
+	Host string
+	// port is the port of the default Endpoint resource.
+	Port uint32
+	// SecLevel controls the security configuration in the Cluster resource.
+	SecLevel SecurityLevel
 }
 
 // DefaultClientResources returns a set of resources (LDS, RDS, CDS, EDS) for a
 // client to generically connect to one server.
-func DefaultClientResources(target, nodeID, host string, port uint32) UpdateOptions {
+func DefaultClientResources(params ResourceParams) UpdateOptions {
 	const routeConfigName = "route"
 	const clusterName = "cluster"
 	const endpointsName = "endpoints"
 
 	return UpdateOptions{
-		NodeID:    nodeID,
-		Listeners: []*v3listenerpb.Listener{DefaultListener(target, routeConfigName)},
-		Routes:    []*v3routepb.RouteConfiguration{DefaultRouteConfig(routeConfigName, target, clusterName)},
-		Clusters:  []*v3clusterpb.Cluster{DefaultCluster(clusterName, endpointsName)},
-		Endpoints: []*v3endpointpb.ClusterLoadAssignment{DefaultEndpoint(endpointsName, host, port)},
+		NodeID:    params.NodeID,
+		Listeners: []*v3listenerpb.Listener{DefaultClientSideListener(params.DialTarget, routeConfigName)},
+		Routes:    []*v3routepb.RouteConfiguration{DefaultRouteConfig(routeConfigName, params.DialTarget, clusterName)},
+		Clusters:  []*v3clusterpb.Cluster{DefaultCluster(clusterName, endpointsName, params.SecLevel)},
+		Endpoints: []*v3endpointpb.ClusterLoadAssignment{DefaultEndpoint(endpointsName, params.Host, params.Port)},
 	}
 }
 
-// DefaultListener returns a basic xds Listener resource.
-func DefaultListener(target, routeName string) *v3listenerpb.Listener {
-	hcm := any(&v3httppb.HttpConnectionManager{
+// DefaultClientSideListener returns a basic xds Listener resource to be used on
+// the client side.
+func DefaultClientSideListener(target, routeName string) *v3listenerpb.Listener {
```

Comment:
Nit: remove "Side" here & "ServerSide" for brevity and to match `DefaultClientResources`.

Or: is there no way to make 1 resource that will work for both?

Reply:
We wont be able to use a single `Listener` for both, because the client listener is expected to have the `apiListener` field set while the server will not.
And as part of our `xdsClient` resource processing, if the `Listener` has an `apiListener`, we assume it to be a client side listener.